File: Sem 3/OOP/Week1_2/bubble.java
// Arrange the elements in ascending and descending order using Bubble sort method. 
import java.util.Scanner;
class Bubble{

	void display(int a[]){
		int n = a.length;
		for(int i = 0; i < n; i++){
			System.out.print(a[i] + " ");
		}
		System.out.println();
	}

	void bubblesort(int a[], boolean ascending){
		int n = a.length;
		for(int i = 0; i < n; i++){
			for(int j = 0; j < n-1-i; j++){
				if(ascending){
					if(a[j] > a[j+1]){
						int temp = a[j];
						a[j] = a[j+1];
						a[j+1] = temp;
					}
				}
				else{
					if(a[j] < a[j+1]){
						int temp = a[j];
						a[j] = a[j+1];
						a[j+1] = temp;
					}
				}
			}
		}
		display(a);
	}
	public static void main(String []args){
		Bubble b = new Bubble();
		Scanner sc = new Scanner(System.in);
		System.out.println("Enter length and elements");
		int n = sc.nextInt();
		int a[] = new int[n];
		for(int i = 0; i < n; i++)
			a[i] = sc.nextInt();
		System.out.println("Ascending: ");
		b.bubblesort(a, true);
		System.out.println("Descending: ");
		b.bubblesort(a, false);
	}
}

File: Sem 4/Algo Lab/Lab3/bubble.c
#include<stdio.h>
#include<stdlib.h>
void bubbleSort(int *a, unsigned int n)
{
    unsigned int i,j,min;
    int temp, flag = 0;
    int opcount=0; // introduce opcount
    for(i= 0; i<n; i++){
        for(j=0; j<n-i-1; j++){
            ++opcount; // increment opcount for every comparison
            if(a[j+1] < a[j]){
                ++opcount;
                flag = 1;
                temp = a[j+1];
                a[j+1] = a[j];
                a[j] = temp;
            }
        }
        if(flag == 0)
            break;
    }
    printf("\nOperation Count %d\n",opcount);
}

int main(){
    int *a;
    int i, n = 5;
    // generate worst case input of different input size
    for (int j=0; j < 6; j++){
        a = (int *)malloc(sizeof(int)*n);
        for (int k=0; k< n; k++)
            a[k] = n-k; // descending order list
        printf("Elements are: ");
        for(i=0;i<n;i++)
            printf("%d ",a[i]);
        bubbleSort(a, n);
        printf("Elements after bubble sort: ");
        for(i = 0; i < n; i++)
            printf("%d ",a[i]);
        printf("\n");
        free(a);
        n = n + 5; // try with a new input size
    }
    return 0;
}